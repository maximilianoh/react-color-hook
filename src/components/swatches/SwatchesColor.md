SwatchesColor without color
```js
import SwatchesColor from './SwatchesColor.jsx';
<SwatchesColor />
```


SwatchesColor color
```js
import SwatchesColor from './SwatchesColor.jsx';
<SwatchesColor color="#56EE54" />
```

SwatchesColor boolean parameter
```js
import SwatchesColor from './SwatchesColor.jsx';
<div>
    <SwatchesColor first color="#56EE54"/>
    <SwatchesColor active color="#EEC954"/>
    <SwatchesColor last color="#5481EE"/>
</div>
```

