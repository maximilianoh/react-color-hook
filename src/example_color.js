export const red = {
  hsl: {
    a: 1, h: 0, l: 0.5, s: 1,
  },
  hex: '#ff0000',
  rgb: {
    r: 255, g: 0, b: 0, a: 1,
  },
  hsv: {
    h: 0, s: 1, v: 1, a: 1,
  },
};

export const yel = {
  hsl: {
    a: 1, h: 50, l: 0.71, s: 0.45,
  },
  hex: '#C4A921',
  rgb: {
    r: 196, g: 169, b: 33, a: 1,
  },
  hsv: {
    h: 50, s: 0.832, v: 0.769, a: 1,
  },
};

export const br = {
  hsl: {
    a: 1, h: 50, l: 0.30, s: 0.45,
  },
  hex: '#6F632A',
  rgb: {
    r: 111, g: 99, b: 42, a: 1,
  },
  hsv: {
    h: 50, s: 0.622, v: 0.435, a: 1,
  },
};
