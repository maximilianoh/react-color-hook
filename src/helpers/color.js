import each from 'lodash/each';
import chroma from 'chroma-js';

const isNumber = (value) => !Number.isNaN(parseFloat(value));

export const simpleCheckForValidColor = (data) => {
  const keysToCheck = ['r', 'g', 'b', 'a', 'h', 's', 'l', 'v'];
  let checked = 0;
  let passed = 0;
  each(keysToCheck, (letter) => {
    if (data[letter]) {
      checked += 1;
      if (isNumber(data[letter])) {
        passed += 1;
      }
      if (letter === 's' || letter === 'l') {
        const percentPatt = /^\d+%$/;
        if (percentPatt.test(data[letter])) {
          passed += 1;
        }
      }
    }
  });
  return (checked === passed) ? data : false;
};

export const chromaValidation = (data) => {
  let validation = false;
  try {
    if (data.hex) chroma(data.hex);
    else chroma(data);
    validation = true;
  } catch (error) {
    validation = false;
  }
  return validation;
};

export const isValidToChange = (data, colors) => {
  const key = data.source;
  if (key === 'rgb') {
    const validation = Number(data.r) === colors.r
    && Number(data.g) === colors.g && Number(data.g) === colors.g;
    return validation;
  }
  if (key === 'hsv') {
    const s = data.s > 1 ? data.s / 100 : data.s;
    const v = data.v > 1 ? data.v / 100 : data.v;
    const validation = Number(data.h) >= 0 && Number(data.h) <= 360
    && Number(s) >= 0 && Number(s) <= 1
    && Number(v) >= 0 && Number(v) <= 1;
    return validation;
  }
  if (key === 'hsl') {
    const s = data.l > 1 ? data.l / 100 : data.l;
    const l = data.l > 1 ? data.l / 100 : data.l;
    try {
      chroma.hsl(data.h, s, l);
      return true;
    } catch (error) {
      return false;
    }
  } else { // hex
    return data.hex.toUpperCase() === colors.toUpperCase();
  }
};

const valueNaN = (h) => (!isNumber(h) ? 0 : h);
const hslaListToObject = (list) => ({
  h: valueNaN(list[0]), s: valueNaN(list[1]), l: valueNaN(list[2]), a: valueNaN(list[3]),
});
const rgbaListToObject = (list) => ({
  r: list[0], g: list[1], b: list[2], a: list[3],
});
const hsvListToObject = (list) => ({
  h: valueNaN(list[0]),
  s: valueNaN(list[1]),
  v: valueNaN(list[2]),
});

export const hsvParse = (color, data) => {
  const s = data.s > 1 ? data.s / 100 : data.s;
  const v = data.v > 1 ? data.v / 100 : data.v;
  const c = chroma.hsv(Number(data.h), Number(s), Number(v));
  const result = {
    ...color,
    hsv: {
      h: Number(data.h),
      s: Number(s),
      v: Number(v),
    },
    hsl: hslaListToObject(c.hsl()),
    rgb: rgbaListToObject(c.rgba()),
    hex: c.hex(),
  };
  return result;
};

export const toState = (data, oldHue) => {
  const dataValidation = data || '#000000';
  let transparent = false;
  let colorChroma;
  let a = data.a ? data.a : 1;
  if (data.hex && (data.hex === 'transparent' || data === 'transparent')) {
    colorChroma = chroma('#000000').alpha(0);
    a = 0;
    transparent = true;
  } else if (data.hex) colorChroma = chroma(data.hex);
  else if (data.h && data.s && data.l) {
    const s = data.s > 1 ? data.s / 100 : data.s;
    const l = data.l > 1 ? data.l / 100 : data.l;
    colorChroma = chroma.hsl(data.h, s, l);
  } else {
    try {
      colorChroma = chroma(dataValidation);
    } catch (error) {
      colorChroma = chroma('#000000');
    }
  }
  const hsl = hslaListToObject(colorChroma.alpha(a).hsl());
  const hsv = hsvListToObject(colorChroma.hsv());
  const rgba = colorChroma.alpha(a).rgba();
  const rgb = rgbaListToObject(rgba);
  const hex = colorChroma.alpha(1).hex();

  return {
    hsl,
    hex: transparent ? 'transparent' : hex,
    rgb,
    hsv,
    oldHue: data.h || oldHue || hsl.h,
    source: data.source,
  };
};

export const isValidHex = (hex) => {
  // disable hex4 and hex8
  // const lh = (String(hex).charAt(0) === '#') ? 1 : 0;
  // const result = hex.length !== (4 + lh) && hex.length < (7 + lh) && chroma.valid(hex);
  // const hasHas = hex.indexOf('#') !== 0 ? '#' : '';

  let validation = false;
  try {
    validation = chroma.valid(hex) && typeof hex === 'string';
  } catch (error) {
    validation = false;
  }
  return validation;
};

export const getContrastingColor = (data) => {
  if (!data) {
    return '#fff';
  }
  let c = data;
  if (data === 'transparent') c = { hex: 'transparent' };
  const col = toState(c);
  if (col.hex === 'transparent') {
    return 'rgba(0,0,0,0.4)';
  }
  const yiq = ((col.rgb.r * 299) + (col.rgb.g * 587) + (col.rgb.b * 114)) / 1000;
  return (yiq >= 128) ? '#000' : '#fff';
};
