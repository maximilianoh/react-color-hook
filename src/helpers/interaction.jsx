import React, { useState } from 'react';

const handleFocus = (Component, Span = 'span') => {
  const Focus = (props) => {
    const [focus, setFocus] = useState(false);

    const handlerFocus = () => setFocus(true);

    const handleBlur = () => setFocus(false);

    return (
      <Span onFocus={handlerFocus} onBlur={handleBlur}>
        <Component {...props} {...focus} />
      </Span>
    );
  };
  return Focus;
};
export default handleFocus;
