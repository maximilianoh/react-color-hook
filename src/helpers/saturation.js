const isNumber = (value) => !Number.isNaN(parseFloat(value));
const calculateChange = (e, hsl, container) => {
  const containerWidth = container.clientWidth;
  const containerHeight = container.clientHeight;
  const x = typeof e.pageX === 'number' ? e.pageX : e.touches[0].pageX;
  const y = typeof e.pageY === 'number' ? e.pageY : e.touches[0].pageY;
  let left = x - (container.getBoundingClientRect().left + window.pageXOffset);
  let top = y - (container.getBoundingClientRect().top + window.pageYOffset);

  if (left < 0) {
    left = 0;
  } else if (left > containerWidth) {
    left = containerWidth;
  }
  if (top < 0) {
    top = 0;
  } else if (top > containerHeight) {
    top = containerHeight;
  }

  const saturation = left / containerWidth;
  const bright = 1 - (top / containerHeight);
  return {
    h: hsl.h,
    s: isNumber(saturation) ? saturation : 0,
    v: isNumber(bright) ? bright : 0,
    a: hsl.a,
    source: 'hsv',
  };
};

export default calculateChange;
