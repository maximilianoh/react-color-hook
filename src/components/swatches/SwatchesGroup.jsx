import React from 'react';
import reactCSS from 'reactcss';
import map from 'lodash/map';
import PropTypes from 'prop-types';
import SwatchesColor from './SwatchesColor';

const SwatchesGroup = ({
  onClick, onSwatchHover, group, active,
}) => {
  const styles = reactCSS({
    default: {
      group: {
        paddingBottom: '10px',
        width: '40px',
        float: 'left',
        marginRight: '10px',
      },
    },
  });

  return (
    <div style={styles.group}>
      {map(group, (color, i) => (
        <SwatchesColor
          key={color}
          color={color}
          active={color.toLowerCase() === active}
          first={i === 0}
          last={i === group.length - 1}
          onClick={onClick}
          onSwatchHover={onSwatchHover}
        />
      ))}
    </div>
  );
};

SwatchesGroup.propTypes = {
  onClick: PropTypes.func,
  onSwatchHover: PropTypes.func,
  group: PropTypes.arrayOf(PropTypes.string).isRequired,
  active: PropTypes.string.isRequired,
};

SwatchesGroup.defaultProps = {
  onClick: () => {},
  onSwatchHover: () => {},
};

export default SwatchesGroup;
