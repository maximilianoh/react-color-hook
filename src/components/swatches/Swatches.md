Swatches
```js
import Swatches from './Swatches.jsx';
<Swatches />
```



Swatches colors
```js
import Swatches from './Swatches.jsx';
import  { red } from '../../helpers/color';
<Swatches hex={ red.hex } colors={ [['#004d40'], ['#333']] } />
```

Swatches custom style
```js
import Swatches from './Swatches.jsx';
import  { red } from '../../helpers/color';
<Swatches hex={ red.hex } styles={{ default: { picker: { boxShadow: '0 0 10px red' } } }} />
```


Swatches Onchange
```js
import Swatches from './Swatches.jsx';
const changeSpy = (props) =>{
    console.dir(props.hex)
};
<div>
    Click and see console
    <Swatches onChange={ changeSpy } />
</div>
```


Swatches onSwatchHover
```js
import Swatches from './Swatches.jsx';
const changeSpy = (props) =>{
    console.dir(props.hex)
};

<div>
    Hover and see console
    <Swatches onSwatchHover={ changeSpy } />
</div>
```