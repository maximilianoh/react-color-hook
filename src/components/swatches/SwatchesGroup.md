SwatchesGroup
```js
import SwatchesGroup from './SwatchesGroup.jsx';
import  { red } from '../../helpers/color';
<SwatchesGroup active={ red.hex } group={ ['#EE6F54', '#56EE54'] } />
```