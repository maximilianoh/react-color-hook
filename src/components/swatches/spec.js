import React from 'react'
import renderer from 'react-test-renderer'
import  { simpleCheckForValidColor } from '../../helpers/color';
import { red } from '../../example_color';
import { render, fireEvent } from '@testing-library/react';
import Swatches from './Swatches'
import SwatchesColor from './SwatchesColor'
import SwatchesGroup from './SwatchesGroup'

test('Swatches renders correctly', () => {
  const tree = renderer.create(
    <Swatches hex={ red.hex } colors={ [['#fff'], ['#333']] } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Swatches renders custom styles correctly', () => {
  const tree = renderer.create(
    <Swatches hex={ red.hex } colors={ [['#fff'], ['#333']] } styles={{ default: { picker: { boxShadow: '0 0 10px red' } } }} />,
  ).toJSON()
  expect(tree.props.style.boxShadow).toBe('0 0 10px red')
})

test('Swatches onChange events correctly', () => {
  const changeSpy = jest.fn((data) => {
    expect(simpleCheckForValidColor(data)).toBeTruthy()
  });

  const {queryByTitle} = render(<Swatches onChange={ changeSpy } />);
  expect(changeSpy).toHaveBeenCalledTimes(0);
  const firstSwatch = queryByTitle("#b71c1c")
  fireEvent.click(firstSwatch);
  expect(changeSpy).toHaveBeenCalled();

})

test('Swatches with onSwatchHover events correctly', () => {
  const hoverSpy = jest.fn((data) => {
    expect(simpleCheckForValidColor(data)).toBeTruthy()
  })

  const {queryByTitle} = render(<Swatches onSwatchHover={ hoverSpy } />);
  expect(hoverSpy).toHaveBeenCalledTimes(0);
  const event = new MouseEvent('mouseover', { bubbles: true });
  const firstSwatch = queryByTitle("#b71c1c")
  fireEvent(firstSwatch, event);
  expect(hoverSpy).toHaveBeenCalled();
})

test('SwatchesColor renders correctly', () => {
  const tree = renderer.create(
    <SwatchesColor />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SwatchesColor renders with props', () => {
  const tree = renderer.create(
    <SwatchesColor active first last />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SwatchesGroup renders correctly', () => {
  const tree = renderer.create(
    <SwatchesGroup active={ red.hex } group={ ['#fff'] } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
