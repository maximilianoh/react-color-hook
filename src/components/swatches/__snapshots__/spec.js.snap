// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swatches renders correctly 1`] = `
<div
  className="swatches-picker "
  style={
    Object {
      "height": 240,
      "width": 320,
    }
  }
>
  <div
    style={
      Object {
        "display": "inline-block",
        "position": "relative",
      }
    }
  >
    <div
      style={
        Object {
          "MozBorderRadius": 2,
          "MozBoxShadow": "0 2px 10px rgba(0,0,0,.12), 0 2px 5px rgba(0,0,0,.16)",
          "OBorderRadius": 2,
          "OBoxShadow": "0 2px 10px rgba(0,0,0,.12), 0 2px 5px rgba(0,0,0,.16)",
          "WebkitBorderRadius": 2,
          "WebkitBoxShadow": "0 2px 10px rgba(0,0,0,.12), 0 2px 5px rgba(0,0,0,.16)",
          "background": "#fff",
          "borderRadius": 2,
          "bottom": "0px",
          "boxShadow": "0 2px 10px rgba(0,0,0,.12), 0 2px 5px rgba(0,0,0,.16)",
          "left": "0px",
          "msBorderRadius": 2,
          "msBoxShadow": "0 2px 10px rgba(0,0,0,.12), 0 2px 5px rgba(0,0,0,.16)",
          "position": "absolute",
          "right": "0px",
          "top": "0px",
        }
      }
    />
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <div
        style={
          Object {
            "height": 240,
            "overflowY": "scroll",
          }
        }
      >
        <div
          style={
            Object {
              "padding": "16px 0 6px 16px",
            }
          }
        >
          <div
            style={
              Object {
                "float": "left",
                "marginRight": "10px",
                "paddingBottom": "10px",
                "width": "40px",
              }
            }
          >
            <span
              onBlur={[Function]}
              onFocus={[Function]}
            >
              <div
                onClick={[Function]}
                onKeyDown={[Function]}
                onMouseOver={[Function]}
                role="button"
                style={
                  Object {
                    "MozBorderRadius": "0 0 2px 2px",
                    "OBorderRadius": "0 0 2px 2px",
                    "WebkitBorderRadius": "0 0 2px 2px",
                    "background": "#fff",
                    "borderRadius": "0 0 2px 2px",
                    "cursor": "pointer",
                    "height": "24px",
                    "marginBottom": "1px",
                    "msBorderRadius": "0 0 2px 2px",
                    "outline": "none",
                    "overflow": "hidden",
                    "position": "relative",
                    "width": "40px",
                  }
                }
                tabIndex={0}
                title="#fff"
              >
                <div
                  style={
                    Object {
                      "color": "#000",
                      "display": "none",
                      "marginLeft": "8px",
                    }
                  }
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                    />
                  </svg>
                </div>
              </div>
            </span>
          </div>
          <div
            style={
              Object {
                "float": "left",
                "marginRight": "10px",
                "paddingBottom": "10px",
                "width": "40px",
              }
            }
          >
            <span
              onBlur={[Function]}
              onFocus={[Function]}
            >
              <div
                onClick={[Function]}
                onKeyDown={[Function]}
                onMouseOver={[Function]}
                role="button"
                style={
                  Object {
                    "MozBorderRadius": "0 0 2px 2px",
                    "OBorderRadius": "0 0 2px 2px",
                    "WebkitBorderRadius": "0 0 2px 2px",
                    "background": "#333",
                    "borderRadius": "0 0 2px 2px",
                    "cursor": "pointer",
                    "height": "24px",
                    "marginBottom": "1px",
                    "msBorderRadius": "0 0 2px 2px",
                    "outline": "none",
                    "overflow": "hidden",
                    "position": "relative",
                    "width": "40px",
                  }
                }
                tabIndex={0}
                title="#333"
              >
                <div
                  style={
                    Object {
                      "color": "#fff",
                      "display": "none",
                      "marginLeft": "8px",
                    }
                  }
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                    />
                  </svg>
                </div>
              </div>
            </span>
          </div>
          <div
            style={
              Object {
                "clear": "both",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SwatchesColor renders correctly 1`] = `
<span
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOver={[Function]}
    role="button"
    style={
      Object {
        "background": "",
        "cursor": "pointer",
        "height": "24px",
        "marginBottom": "1px",
        "outline": "none",
        "position": "relative",
        "width": "40px",
      }
    }
    tabIndex={0}
    title=""
  >
    <div
      style={
        Object {
          "color": "#fff",
          "display": "none",
          "marginLeft": "8px",
        }
      }
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
        />
      </svg>
    </div>
  </div>
</span>
`;

exports[`SwatchesColor renders with props 1`] = `
<span
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOver={[Function]}
    role="button"
    style={
      Object {
        "MozBorderRadius": "0 0 2px 2px",
        "OBorderRadius": "0 0 2px 2px",
        "WebkitBorderRadius": "0 0 2px 2px",
        "background": "",
        "borderRadius": "0 0 2px 2px",
        "cursor": "pointer",
        "height": "24px",
        "marginBottom": "1px",
        "msBorderRadius": "0 0 2px 2px",
        "outline": "none",
        "overflow": "hidden",
        "position": "relative",
        "width": "40px",
      }
    }
    tabIndex={0}
    title=""
  >
    <div
      style={
        Object {
          "color": "#fff",
          "display": "block",
          "marginLeft": "8px",
        }
      }
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
        />
      </svg>
    </div>
  </div>
</span>
`;

exports[`SwatchesGroup renders correctly 1`] = `
<div
  style={
    Object {
      "float": "left",
      "marginRight": "10px",
      "paddingBottom": "10px",
      "width": "40px",
    }
  }
>
  <span
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      onClick={[Function]}
      onKeyDown={[Function]}
      onMouseOver={[Function]}
      role="button"
      style={
        Object {
          "MozBorderRadius": "0 0 2px 2px",
          "OBorderRadius": "0 0 2px 2px",
          "WebkitBorderRadius": "0 0 2px 2px",
          "background": "#fff",
          "borderRadius": "0 0 2px 2px",
          "cursor": "pointer",
          "height": "24px",
          "marginBottom": "1px",
          "msBorderRadius": "0 0 2px 2px",
          "outline": "none",
          "overflow": "hidden",
          "position": "relative",
          "width": "40px",
        }
      }
      tabIndex={0}
      title="#fff"
    >
      <div
        style={
          Object {
            "color": "#000",
            "display": "none",
            "marginLeft": "8px",
          }
        }
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
      </div>
    </div>
  </span>
</div>
`;
